impl/src/main/java/org/ehcache/internal/store/heap/OnHeapStore.java

```diff
@@ -401,26 +491,34 @@ private void invalidate() {
       
       private void advance() {
         next = null;
-        while (next == null && it.hasNext()) {
-          Map.Entry<K, OnHeapValueHolder<V>> entry = it.next();
-          final long now = timeSource.getTimeMillis();
-          if (entry.getValue().isExpired(now, TimeUnit.MILLISECONDS)) {
-            it.remove();
-            onExpiration(entry.getKey(), entry.getValue());
-            continue;
+        try {
+          while (next == null && it.hasNext()) {
+            Map.Entry<K, OnHeapValueHolder<V>> entry = it.next();
+            final long now = timeSource.getTimeMillis();
+            if (entry.getValue().isExpired(now, TimeUnit.MILLISECONDS)) {
+              it.remove();
+              onExpiration(entry.getKey(), entry.getValue());
+              continue;
+            }
+
+            next = entry;
           }
-          
-          next = entry;
+        } catch (RuntimeException re) {
+          prefetchFailure = new CacheAccessException(re);
         }
       }
       
       @Override
-      public boolean hasNext() throws CacheAccessException {
+      public boolean hasNext() {
         return next != null;
```

Comment:
But that is wrong, no? _hasNext_ is supposed to return whether or not a next element is there. A prefetchFailure happens when a next element "is there" but we just failed to retrieve that. So _hasNext_ returning false in that case is like lying to the user, right? The next element is actually there. Even though there is no value in returning *true* either since that is not useful to the user in any way.